{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare African allele frequency data on variants of interest to that of other global populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules and packages\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from upsetplot import plot\n",
    "from requests import get, codes as http_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set constants and functions\n",
    "\n",
    "home_path = str(os.path.dirname(os.getcwd()))\n",
    "population_clusters = [\"SUPER\",\"SUB\"]\n",
    "\n",
    "gene_location_df = pd.read_csv(os.path.join(home_path, \"Data_descriptions\", \"locations.csv\"))\n",
    "genes = gene_location_df.location_name\n",
    "\n",
    "sample_info_df = pd.read_csv(os.path.join(home_path, \"Data_descriptions\", \"samples.csv\"))\n",
    "sub_populations = sample_info_df.SUB.unique()\n",
    "regional_classification = {\"ACB\":\"ACB\",\"ASW\":\"ASW\",\"GWD\":\"WA\",\"ESN\":\"WA\",\"MSL\":\"WA\",\"MbutiPygmy\":\"CA\",\"BiakaPygmy\":\"CA\",\"Mandenka\":\"WA\",\"Yoruba\":\"WA\",\"San\":\"KS\",\"BantuSouthAfrica\":\"SA\",\"BantuKenya\":\"EA\",\"YRI\":\"WA\",\"LWK\":\"EA\"} # SA = South Africa, CA = Central Africa, WA = West Africa, EA = East Africa, ACB = African Carribean in Barbados, ASW = African American, KS = Khoi-San\n",
    "\n",
    "# Create a list with variant IDs of interest\n",
    "NESHIE_rsIDs = [\"rs2067853\",\"rs1217401\",\"rs2043211\",\"rs1001179\",\"rs1961495\",\"rs1411040\",\"rs34004222\",\"rs13027659\",\"rs190148408\",\"rs79704487\",\"rs1800896\",\"rs3024490\",\"rs1800871\",\"rs1554286\",\"rs1518111\",\"rs1143623\",\"rs16944\",\"rs1071676\",\"rs1800795\",\"rs2069837\",\"rs1800796\",\"rs2066992\",\"rs2069832\",\"rs2069833\",\"rs1554606\",\"rs2069845\",\"rs1801133\",\"rs4846049\",\"rs1476413\",\"rs1801131\",\"rs9651118\",\"rs1808593\",\"rs2070744\",\"rs1800779\",\"rs6517135\",\"rs1799964\",\"rs1799724\",\"rs361525\",\"rs1800629\"]\n",
    "\n",
    "# Set plot figure area and font size\n",
    "sns.set(rc={\"figure.figsize\": (12, 10)})\n",
    "SMALL_font = 8\n",
    "MEDIUM_font = 10\n",
    "BIGGER_font = 16\n",
    "\n",
    "plt.rc('font', size=SMALL_font)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=SMALL_font)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_font)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=SMALL_font)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=SMALL_font)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=SMALL_font)    # legend fontsize\n",
    "plt.rc('figure', titlesize=BIGGER_font)  # fontsize of the figure title\n",
    "\n",
    "# Functions for retrieving variant allele ALFA count information from the NCBI API\n",
    "\n",
    "def get_ALFA_count_info(variant_id_number: str):\n",
    "  \"\"\"\n",
    "  Retrieve ALFA allele counts from NCBI in JSON format\n",
    "  \"\"\"\n",
    "  url = \"https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/{}/frequency\".format(variant_id_number)\n",
    "  request = get(url)\n",
    "  if request.status_code != http_code.ok:\n",
    "    raise Exception(\"Request failed: {}\".format(request.status_code))\n",
    "  else:\n",
    "    request_json = request.json()\n",
    "  return request_json\n",
    "\n",
    "def add_all_pops(populations, project):\n",
    "  \"\"\"\n",
    "  Add metadata to translate ALFA population code data into legible population names into a single dataframe. Code copied from https://github.com/ncbi/dbsnp/blob/master/tutorials/Variation%20Services/Jupyter_Notebook/by_rsid.ipynb.\n",
    "  \"\"\"  \n",
    "  for p in populations:\n",
    "    project['pops'][p['biosample_id']] = p\n",
    "  if 'subs' in p:\n",
    "    add_all_pops(p['subs'], project)\n",
    "\n",
    "def allele_counts (study_id, study_counts):\n",
    "  \"\"\"\n",
    "  Parse study count information and return reference and alternate allele counts. Code modified from https://github.com/ncbi/dbsnp/blob/master/tutorials/Variation%20Services/Jupyter_Notebook/by_rsid.ipynb. \n",
    "  \"\"\"\n",
    "  allele_counts = study_counts[\"allele_counts\"]\n",
    "  pop_allele_counts = pd.DataFrame()\n",
    "  for pop_id, pop_counts in allele_counts.items():\n",
    "      for allele, count in pop_counts.items():\n",
    "          study_id = study_id\n",
    "          pop_id = pop_id\n",
    "          return (\"Allele: {}, Count: {}\".format(allele, count))\n",
    "\n",
    "def get_metadata():\n",
    "  \"\"\"\n",
    "  Retrieve ALFA allele count metadata \n",
    "  \"\"\"\n",
    "  url = \"https://api.ncbi.nlm.nih.gov/variation/v0/metadata/frequency\"\n",
    "  request = get(url)\n",
    "  if request.status_code != http_code.ok:\n",
    "    raise Exception(\"Request failed: {}\".format(request.status_code))\n",
    "  else:\n",
    "    request_json = request.json()\n",
    "  return request_json\n",
    "\n",
    "def calculate_frequencies(ref_count: int, alt_count: int):\n",
    "  \"\"\"\n",
    "  Calculate allele frequencies from reference and alternate allele count information\n",
    "  \"\"\"\n",
    "  total_count = ref_count + alt_count\n",
    "  alt_frequency = alt_count / total_count\n",
    "  ref_frequency = ref_count / total_count\n",
    "\n",
    "  return (alt_frequency, ref_frequency)\n",
    "\n",
    "def add_prefix_dataframe_col_names(dataframe, columns_list, prefix):\n",
    "    rename_dict = dict()\n",
    "    for line in columns_list:\n",
    "        key = line\n",
    "        if key not in rename_dict:\n",
    "            rename_dict[key] = str()\n",
    "        rename_dict[key] = prefix + \"{}\".format(line)\n",
    "    dataframe = dataframe.rename(columns = rename_dict)\n",
    "    return dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract global allele count data on NESHIE variant IDs from the NCBI ALFA Allele Frequency Project via API.\n",
    "\n",
    "variant_global_allele_counts = pd.DataFrame()\n",
    "\n",
    "for variant_id in NESHIE_rsIDs:\n",
    "    variant_id_number = variant_id.replace(\"rs\",\"\")\n",
    "    count_data = get_ALFA_count_info(variant_id_number)\n",
    "    population_count_data = count_data[\"results\"]\n",
    "    for interval, data in population_count_data.items():\n",
    "      variant_ref = data['ref']\n",
    "      variant_study_counts = data[\"counts\"]\n",
    "      for study_code, study_allele_counts in variant_study_counts.items():\n",
    "        population_counts = study_allele_counts[\"allele_counts\"]\n",
    "        variant_population_allele_count = pd.DataFrame()\n",
    "        for population_code, allele_counts in population_counts.items():\n",
    "          variant_population_allele_count[\"study_code\"] = [study_code]            \n",
    "          variant_population_allele_count[\"variant_id\"] = [variant_id]\n",
    "          variant_population_allele_count[\"population_code\"] = [population_code]\n",
    "          variant_population_allele_count[\"reference_allele\"] = [variant_ref]\n",
    "          variant_population_allele_count[\"allele_counts\"] = [allele_counts]\n",
    "          variant_global_allele_counts = pd.concat([variant_global_allele_counts, variant_population_allele_count])       \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_code</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>population_code</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492705</td>\n",
       "      <td>G</td>\n",
       "      <td>42603.0</td>\n",
       "      <td>101389.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492695</td>\n",
       "      <td>G</td>\n",
       "      <td>39304.0</td>\n",
       "      <td>84208.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492703</td>\n",
       "      <td>G</td>\n",
       "      <td>434.0</td>\n",
       "      <td>6158.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492696</td>\n",
       "      <td>G</td>\n",
       "      <td>8.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492698</td>\n",
       "      <td>G</td>\n",
       "      <td>426.0</td>\n",
       "      <td>5936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492701</td>\n",
       "      <td>G</td>\n",
       "      <td>121.0</td>\n",
       "      <td>1737.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492699</td>\n",
       "      <td>G</td>\n",
       "      <td>73.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492700</td>\n",
       "      <td>G</td>\n",
       "      <td>455.0</td>\n",
       "      <td>4911.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492702</td>\n",
       "      <td>G</td>\n",
       "      <td>30.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN11605645</td>\n",
       "      <td>G</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>12967.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>468 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      study_code variant_id population_code reference_allele        A  \\\n",
       "0    PRJNA507278  rs2067853    SAMN10492705                G  42603.0   \n",
       "1    PRJNA507278  rs2067853    SAMN10492695                G  39304.0   \n",
       "2    PRJNA507278  rs2067853    SAMN10492703                G    434.0   \n",
       "3    PRJNA507278  rs2067853    SAMN10492696                G      8.0   \n",
       "4    PRJNA507278  rs2067853    SAMN10492698                G    426.0   \n",
       "..           ...        ...             ...              ...      ...   \n",
       "463  PRJNA507278  rs1800629    SAMN10492701                G    121.0   \n",
       "464  PRJNA507278  rs1800629    SAMN10492699                G     73.0   \n",
       "465  PRJNA507278  rs1800629    SAMN10492700                G    455.0   \n",
       "466  PRJNA507278  rs1800629    SAMN10492702                G     30.0   \n",
       "467  PRJNA507278  rs1800629    SAMN11605645                G   1821.0   \n",
       "\n",
       "            G   T   C  \n",
       "0    101389.0 NaN NaN  \n",
       "1     84208.0 NaN NaN  \n",
       "2      6158.0 NaN NaN  \n",
       "3       222.0 NaN NaN  \n",
       "4      5936.0 NaN NaN  \n",
       "..        ...  ..  ..  \n",
       "463    1737.0 NaN NaN  \n",
       "464     511.0 NaN NaN  \n",
       "465    4911.0 NaN NaN  \n",
       "466     294.0 NaN NaN  \n",
       "467   12967.0 NaN NaN  \n",
       "\n",
       "[468 rows x 8 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Normalise allele count column of variant count dataframe\n",
    "normalised_count_data = pd.json_normalize(variant_global_allele_counts.allele_counts)\n",
    "drop_count_data = variant_global_allele_counts.drop(columns=\"allele_counts\")\n",
    "normalised_variant_count_data = pd.concat([drop_count_data.reset_index(drop=True), normalised_count_data.reset_index(drop=True)], axis = 1) \n",
    "normalised_variant_count_data  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_code</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>population_code</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492705</td>\n",
       "      <td>G</td>\n",
       "      <td>42603.0</td>\n",
       "      <td>101389.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Total</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492695</td>\n",
       "      <td>G</td>\n",
       "      <td>39304.0</td>\n",
       "      <td>84208.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>European</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492703</td>\n",
       "      <td>G</td>\n",
       "      <td>434.0</td>\n",
       "      <td>6158.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492696</td>\n",
       "      <td>G</td>\n",
       "      <td>8.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>African Others</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs2067853</td>\n",
       "      <td>SAMN10492698</td>\n",
       "      <td>G</td>\n",
       "      <td>426.0</td>\n",
       "      <td>5936.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>African American</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492701</td>\n",
       "      <td>G</td>\n",
       "      <td>121.0</td>\n",
       "      <td>1737.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other Asian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492699</td>\n",
       "      <td>G</td>\n",
       "      <td>73.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Latin American 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492700</td>\n",
       "      <td>G</td>\n",
       "      <td>455.0</td>\n",
       "      <td>4911.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Latin American 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN10492702</td>\n",
       "      <td>G</td>\n",
       "      <td>30.0</td>\n",
       "      <td>294.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>South Asian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>PRJNA507278</td>\n",
       "      <td>rs1800629</td>\n",
       "      <td>SAMN11605645</td>\n",
       "      <td>G</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>12967.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>468 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      study_code variant_id population_code reference_allele        A  \\\n",
       "0    PRJNA507278  rs2067853    SAMN10492705                G  42603.0   \n",
       "1    PRJNA507278  rs2067853    SAMN10492695                G  39304.0   \n",
       "2    PRJNA507278  rs2067853    SAMN10492703                G    434.0   \n",
       "3    PRJNA507278  rs2067853    SAMN10492696                G      8.0   \n",
       "4    PRJNA507278  rs2067853    SAMN10492698                G    426.0   \n",
       "..           ...        ...             ...              ...      ...   \n",
       "463  PRJNA507278  rs1800629    SAMN10492701                G    121.0   \n",
       "464  PRJNA507278  rs1800629    SAMN10492699                G     73.0   \n",
       "465  PRJNA507278  rs1800629    SAMN10492700                G    455.0   \n",
       "466  PRJNA507278  rs1800629    SAMN10492702                G     30.0   \n",
       "467  PRJNA507278  rs1800629    SAMN11605645                G   1821.0   \n",
       "\n",
       "            G   T   C        Population  \n",
       "0    101389.0 NaN NaN             Total  \n",
       "1     84208.0 NaN NaN          European  \n",
       "2      6158.0 NaN NaN           African  \n",
       "3       222.0 NaN NaN    African Others  \n",
       "4      5936.0 NaN NaN  African American  \n",
       "..        ...  ..  ..               ...  \n",
       "463    1737.0 NaN NaN       Other Asian  \n",
       "464     511.0 NaN NaN  Latin American 1  \n",
       "465    4911.0 NaN NaN  Latin American 2  \n",
       "466     294.0 NaN NaN       South Asian  \n",
       "467   12967.0 NaN NaN             Other  \n",
       "\n",
       "[468 rows x 9 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate population codes into legible population names.\n",
    "\n",
    "# Fetch study population metadata \n",
    "metadata_json = get_metadata()\n",
    "\n",
    "# Generate list of all unique study and population codes that need to be queried\n",
    "\n",
    "study_codes = []\n",
    "for study_code in normalised_variant_count_data.study_code.unique(): \n",
    "    study_codes.append(study_code)\n",
    "\n",
    "population_codes = []\n",
    "for population_code in normalised_variant_count_data.population_code.unique():\n",
    "    population_codes.append(population_code)\n",
    "\n",
    "# Fetch study and population name metadata. Code modified from https://github.com/ncbi/dbsnp/blob/master/tutorials/Variation%20Services/Jupyter_Notebook/by_rsid.ipynb.\n",
    "metadata = {}\n",
    "for project_json in metadata_json:\n",
    "  p = {}\n",
    "  p['json']=project_json\n",
    "  p['pops']={}\n",
    "  metadata[project_json['bioproject_id']] = p\n",
    "for prj_id, prj in metadata.items():\n",
    "  add_all_pops(prj['json']['populations'], prj)\n",
    "\n",
    "for study_id in study_codes:\n",
    "    study_name = metadata[study_id]['json']['short_name']\n",
    "\n",
    "    pop_dict = {\"SAMN10492696\" : \"African Others\", \"SAMN10492698\" : \"African American\", \"SAMN10492697\" : \"East Asian\", \"SAMN10492701\" : \"Other Asian\"}\n",
    "    for pop_id in population_codes: \n",
    "        if pop_id not in [\"SAMN10492696\", \"SAMN10492698\", \"SAMN10492697\", \"SAMN10492701\"]:\n",
    "            population_name = metadata[study_id]['pops'][pop_id]['name']\n",
    "            temp_pop_dict = {pop_id : population_name}\n",
    "        pop_dict.update(temp_pop_dict)\n",
    "\n",
    "# Add column in population allele count dataframe for population names\n",
    "\n",
    "normalised_variant_count_data[\"Population\"] = normalised_variant_count_data[\"population_code\"].map(pop_dict)\n",
    "normalised_variant_count_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>Population</th>\n",
       "      <th>G_freq</th>\n",
       "      <th>C_freq</th>\n",
       "      <th>A_freq</th>\n",
       "      <th>T_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>Total</td>\n",
       "      <td>0.704129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295871</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>European</td>\n",
       "      <td>0.681780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.318220</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>African</td>\n",
       "      <td>0.934163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.065837</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>African Others</td>\n",
       "      <td>0.965217</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.034783</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>African American</td>\n",
       "      <td>0.933040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.066960</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>G</td>\n",
       "      <td>Other Asian</td>\n",
       "      <td>0.934876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.065124</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>G</td>\n",
       "      <td>Latin American 1</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>G</td>\n",
       "      <td>Latin American 2</td>\n",
       "      <td>0.915207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.084793</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>G</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>0.907407</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>G</td>\n",
       "      <td>Other</td>\n",
       "      <td>0.876860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.123140</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>468 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id reference_allele        Population    G_freq  C_freq    A_freq  \\\n",
       "0    rs2067853                G             Total  0.704129     NaN  0.295871   \n",
       "1    rs2067853                G          European  0.681780     NaN  0.318220   \n",
       "2    rs2067853                G           African  0.934163     NaN  0.065837   \n",
       "3    rs2067853                G    African Others  0.965217     NaN  0.034783   \n",
       "4    rs2067853                G  African American  0.933040     NaN  0.066960   \n",
       "..         ...              ...               ...       ...     ...       ...   \n",
       "463  rs1800629                G       Other Asian  0.934876     NaN  0.065124   \n",
       "464  rs1800629                G  Latin American 1  0.875000     NaN  0.125000   \n",
       "465  rs1800629                G  Latin American 2  0.915207     NaN  0.084793   \n",
       "466  rs1800629                G       South Asian  0.907407     NaN  0.092593   \n",
       "467  rs1800629                G             Other  0.876860     NaN  0.123140   \n",
       "\n",
       "     T_freq  \n",
       "0       NaN  \n",
       "1       NaN  \n",
       "2       NaN  \n",
       "3       NaN  \n",
       "4       NaN  \n",
       "..      ...  \n",
       "463     NaN  \n",
       "464     NaN  \n",
       "465     NaN  \n",
       "466     NaN  \n",
       "467     NaN  \n",
       "\n",
       "[468 rows x 7 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate allele frequencies from allele count data and append to allele count dataframe\n",
    "normalised_variant_count_data[\"total_count\"] = normalised_variant_count_data[[\"G\", \"C\", \"A\", \"T\"]].sum(axis=1)\n",
    "\n",
    "normalised_variant_count_data[\"G_freq\"] = normalised_variant_count_data[\"G\"] / normalised_variant_count_data[\"total_count\"]\n",
    "normalised_variant_count_data[\"C_freq\"] = normalised_variant_count_data[\"C\"] / normalised_variant_count_data[\"total_count\"]\n",
    "normalised_variant_count_data[\"A_freq\"] = normalised_variant_count_data[\"A\"] / normalised_variant_count_data[\"total_count\"]\n",
    "normalised_variant_count_data[\"T_freq\"] = normalised_variant_count_data[\"T\"] / normalised_variant_count_data[\"total_count\"]\n",
    "\n",
    "ALFA_freq_data = normalised_variant_count_data.drop(columns=[\"G\", \"C\", \"A\", \"T\", \"total_count\", \"study_code\", \"population_code\"])\n",
    "ALFA_freq_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_21596\\2094901536.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  collated_count_subset[\"REG\"] = collated_count_subset[\"SUB_POP\"].map(regional_classification)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_CTS</th>\n",
       "      <th>REF_CTS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3716</th>\n",
       "      <td>13</td>\n",
       "      <td>rs34004222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>188</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4941</th>\n",
       "      <td>13</td>\n",
       "      <td>rs1961495</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>34</td>\n",
       "      <td>154</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8871</th>\n",
       "      <td>13</td>\n",
       "      <td>rs1411040</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>161</td>\n",
       "      <td>27</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3716</th>\n",
       "      <td>13</td>\n",
       "      <td>rs34004222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>232</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>GWD</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4941</th>\n",
       "      <td>13</td>\n",
       "      <td>rs1961495</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>15</td>\n",
       "      <td>217</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>GWD</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>1</td>\n",
       "      <td>rs4846049</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>ASW</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1476413</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>16</td>\n",
       "      <td>88</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>ASW</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>84</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>ASW</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1801133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>88</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>ASW</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>1</td>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>100</td>\n",
       "      <td>MTHFR</td>\n",
       "      <td>ASW</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #CHROM          ID REF ALT  ALT_CTS  REF_CTS    GENE SUB_POP  REG\n",
       "3716      13  rs34004222   G   A        0      188  COL4A1     ACB  ACB\n",
       "4941      13   rs1961495   C   T       34      154  COL4A1     ACB  ACB\n",
       "8871      13   rs1411040   C   T      161       27  COL4A1     ACB  ACB\n",
       "3716      13  rs34004222   G   A        0      232  COL4A1     GWD   WA\n",
       "4941      13   rs1961495   C   T       15      217  COL4A1     GWD   WA\n",
       "...      ...         ...  ..  ..      ...      ...     ...     ...  ...\n",
       "259        1   rs4846049   T   G       51       53   MTHFR     ASW  ASW\n",
       "382        1   rs1476413   C   T       16       88   MTHFR     ASW  ASW\n",
       "542        1   rs1801131   T   G       20       84   MTHFR     ASW  ASW\n",
       "623        1   rs1801133   G   A       16       88   MTHFR     ASW  ASW\n",
       "944        1   rs9651118   T   C        4      100   MTHFR     ASW  ASW\n",
       "\n",
       "[406 rows x 9 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import in-house allele count data\n",
    "\n",
    "collated_count_data = pd.DataFrame()\n",
    "for gene in genes:\n",
    "    for sub_population in sub_populations:\n",
    "        allele_count_path = os.path.join(home_path,\"Data\", \"SUB\", \"ALL_{}.{}.acount\".format(gene, sub_population))\n",
    "        if os.path.exists(allele_count_path):\n",
    "            allele_count_df = pd.read_csv(allele_count_path, sep=\"\\t\")\n",
    "            allele_count_df[\"REF_CTS\"] = allele_count_df[\"OBS_CT\"] - allele_count_df[\"ALT_CTS\"]\n",
    "            allele_count_df[\"GENE\"] = gene\n",
    "            allele_count_df[\"SUB_POP\"] = sub_population\n",
    "        collated_count_data = pd.concat([collated_count_data, allele_count_df]).drop(columns=\"OBS_CT\")\n",
    "\n",
    "# Filter in-house count data for variants of interest\n",
    "collated_count_subset = collated_count_data[collated_count_data.ID.isin(NESHIE_rsIDs)]\n",
    "collated_count_subset[\"REG\"] = collated_count_subset[\"SUB_POP\"].map(regional_classification)\n",
    "collated_count_subset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_CTS</th>\n",
       "      <th>REF_CTS</th>\n",
       "      <th>REG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>KS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>SA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>212</td>\n",
       "      <td>1298</td>\n",
       "      <td>African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>26</td>\n",
       "      <td>162</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>173</td>\n",
       "      <td>1045</td>\n",
       "      <td>Native African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1218</td>\n",
       "      <td>Native African</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>166</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>206</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1206</td>\n",
       "      <td>Native African without KS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>522 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID REF ALT  ALT_CTS  REF_CTS                        REG\n",
       "0    rs1071676   C   G        4        8                         KS\n",
       "1    rs1071676   C   G        2       14                         SA\n",
       "2    rs1071676   C   G      212     1298                    African\n",
       "3    rs1071676   C   G       26      162                        ACB\n",
       "4    rs1071676   C   G      173     1045             Native African\n",
       "..         ...  ..  ..      ...      ...                        ...\n",
       "517  rs9651118   T   C        0     1218             Native African\n",
       "518  rs9651118   T   C        0      166                         WA\n",
       "519  rs9651118   T   C        0      206                         WA\n",
       "520  rs9651118   T   C        0       42                         WA\n",
       "521  rs9651118   T   C        0     1206  Native African without KS\n",
       "\n",
       "[522 rows x 6 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate in-house data allele counts for the total African population, the total African population excluding non-native African American and Carribean Africans, the total African population excluding Khoi-San people, and Southern/South Africa with the addition of Khoi-San populations. Append the additional allele frequency data to the existing allele frequency dataframe. \n",
    "collated_count_subset_b = collated_count_subset[[\"ID\",\"REF\",\"ALT\",\"ALT_CTS\",\"REF_CTS\",\"REG\"]].copy()\n",
    "total_africa_ct = collated_count_subset_b.groupby([\"ID\", \"REF\", \"ALT\"]).sum(numeric_only=True).reset_index().assign(REG = 'African')\n",
    "native_africa_ct = collated_count_subset_b[(collated_count_subset.REG != \"ACB\") & (collated_count_subset.REG != \"ASW\")].groupby([\"ID\", \"REF\", \"ALT\"]).sum(numeric_only=True).reset_index().assign(REG = 'Native African')\n",
    "native_africa_no_ks_ct = collated_count_subset_b[(collated_count_subset.REG != \"KS\") & (collated_count_subset.REG != \"ACB\") & (collated_count_subset.REG != \"ASW\")].groupby([\"ID\", \"REF\", \"ALT\"]).sum(numeric_only=True).reset_index().assign(REG = 'Native African without KS')\n",
    "SA_KS_ct = collated_count_subset_b[(collated_count_subset.REG == \"SA\") | (collated_count_subset.REG == \"KS\")].groupby([\"ID\", \"REF\", \"ALT\"]).sum(numeric_only=True).reset_index().assign(REG = 'SA and KS')\n",
    "\n",
    "collated_count_data = pd.concat([collated_count_subset_b, total_africa_ct, native_africa_ct, SA_KS_ct, native_africa_no_ks_ct]).sort_values(\"ID\").reset_index(drop=True)\n",
    "collated_count_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SA',\n",
       " 'Native African',\n",
       " 'ASW',\n",
       " 'KS',\n",
       " 'ACB',\n",
       " 'African',\n",
       " 'SA and KS',\n",
       " 'WA',\n",
       " 'Native African without KS',\n",
       " 'EA',\n",
       " 'CA']"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate list of all unique regions in collated_count_data dataframe\n",
    "\n",
    "inhouse_populations = []\n",
    "for region in set(collated_count_data.REG.values):\n",
    "    inhouse_populations.append(region)\n",
    "\n",
    "inhouse_populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>REG</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>IH_ALT_CT_ACB</th>\n",
       "      <th>IH_ALT_CT_ASW</th>\n",
       "      <th>IH_ALT_CT_African</th>\n",
       "      <th>IH_ALT_CT_CA</th>\n",
       "      <th>IH_ALT_CT_EA</th>\n",
       "      <th>IH_ALT_CT_KS</th>\n",
       "      <th>IH_ALT_CT_Native African</th>\n",
       "      <th>...</th>\n",
       "      <th>IH_REF_CT_ASW</th>\n",
       "      <th>IH_REF_CT_African</th>\n",
       "      <th>IH_REF_CT_CA</th>\n",
       "      <th>IH_REF_CT_EA</th>\n",
       "      <th>IH_REF_CT_KS</th>\n",
       "      <th>IH_REF_CT_Native African</th>\n",
       "      <th>IH_REF_CT_Native African without KS</th>\n",
       "      <th>IH_REF_CT_SA</th>\n",
       "      <th>IH_REF_CT_SA and KS</th>\n",
       "      <th>IH_REF_CT_WA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1071676</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>162</td>\n",
       "      <td>91</td>\n",
       "      <td>1298</td>\n",
       "      <td>22</td>\n",
       "      <td>163</td>\n",
       "      <td>8</td>\n",
       "      <td>1045</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1298</td>\n",
       "      <td>22</td>\n",
       "      <td>163</td>\n",
       "      <td>8</td>\n",
       "      <td>1045</td>\n",
       "      <td>1037</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1217401</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>45</td>\n",
       "      <td>25</td>\n",
       "      <td>317</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>247</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>317</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>247</td>\n",
       "      <td>243</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1411040</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "      <td>264</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>224</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>264</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>224</td>\n",
       "      <td>223</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1476413</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>154</td>\n",
       "      <td>88</td>\n",
       "      <td>1287</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>1045</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>1287</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>1045</td>\n",
       "      <td>1037</td>\n",
       "      <td>12</td>\n",
       "      <td>20</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1518111</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>71</td>\n",
       "      <td>38</td>\n",
       "      <td>650</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>541</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>650</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>541</td>\n",
       "      <td>540</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1554286</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>69</td>\n",
       "      <td>38</td>\n",
       "      <td>646</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>539</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>646</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>539</td>\n",
       "      <td>538</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1554606</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>67</td>\n",
       "      <td>34</td>\n",
       "      <td>473</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>372</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>473</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>372</td>\n",
       "      <td>366</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1800779</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>217</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>217</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>176</td>\n",
       "      <td>176</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1800795</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1800796</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>171</td>\n",
       "      <td>97</td>\n",
       "      <td>1354</td>\n",
       "      <td>42</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>1086</td>\n",
       "      <td>...</td>\n",
       "      <td>97</td>\n",
       "      <td>1354</td>\n",
       "      <td>42</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>1086</td>\n",
       "      <td>1075</td>\n",
       "      <td>14</td>\n",
       "      <td>25</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1800871</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>72</td>\n",
       "      <td>38</td>\n",
       "      <td>653</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>653</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>542</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1800896</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>131</td>\n",
       "      <td>63</td>\n",
       "      <td>1016</td>\n",
       "      <td>6</td>\n",
       "      <td>124</td>\n",
       "      <td>7</td>\n",
       "      <td>822</td>\n",
       "      <td>...</td>\n",
       "      <td>63</td>\n",
       "      <td>1016</td>\n",
       "      <td>6</td>\n",
       "      <td>124</td>\n",
       "      <td>7</td>\n",
       "      <td>822</td>\n",
       "      <td>815</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1801131</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>153</td>\n",
       "      <td>84</td>\n",
       "      <td>1277</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>1040</td>\n",
       "      <td>...</td>\n",
       "      <td>84</td>\n",
       "      <td>1277</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>1040</td>\n",
       "      <td>1032</td>\n",
       "      <td>12</td>\n",
       "      <td>20</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1801133</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>166</td>\n",
       "      <td>88</td>\n",
       "      <td>1384</td>\n",
       "      <td>42</td>\n",
       "      <td>173</td>\n",
       "      <td>12</td>\n",
       "      <td>1130</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>1384</td>\n",
       "      <td>42</td>\n",
       "      <td>173</td>\n",
       "      <td>12</td>\n",
       "      <td>1130</td>\n",
       "      <td>1118</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1808593</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>43</td>\n",
       "      <td>26</td>\n",
       "      <td>430</td>\n",
       "      <td>3</td>\n",
       "      <td>60</td>\n",
       "      <td>10</td>\n",
       "      <td>361</td>\n",
       "      <td>...</td>\n",
       "      <td>26</td>\n",
       "      <td>430</td>\n",
       "      <td>3</td>\n",
       "      <td>60</td>\n",
       "      <td>10</td>\n",
       "      <td>361</td>\n",
       "      <td>351</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs1961495</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>154</td>\n",
       "      <td>93</td>\n",
       "      <td>1319</td>\n",
       "      <td>23</td>\n",
       "      <td>156</td>\n",
       "      <td>12</td>\n",
       "      <td>1072</td>\n",
       "      <td>...</td>\n",
       "      <td>93</td>\n",
       "      <td>1319</td>\n",
       "      <td>23</td>\n",
       "      <td>156</td>\n",
       "      <td>12</td>\n",
       "      <td>1072</td>\n",
       "      <td>1060</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs2043211</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>151</td>\n",
       "      <td>81</td>\n",
       "      <td>1232</td>\n",
       "      <td>35</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>1000</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>1232</td>\n",
       "      <td>35</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>1000</td>\n",
       "      <td>989</td>\n",
       "      <td>14</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs2066992</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>176</td>\n",
       "      <td>97</td>\n",
       "      <td>1382</td>\n",
       "      <td>24</td>\n",
       "      <td>145</td>\n",
       "      <td>12</td>\n",
       "      <td>1109</td>\n",
       "      <td>...</td>\n",
       "      <td>97</td>\n",
       "      <td>1382</td>\n",
       "      <td>24</td>\n",
       "      <td>145</td>\n",
       "      <td>12</td>\n",
       "      <td>1109</td>\n",
       "      <td>1097</td>\n",
       "      <td>14</td>\n",
       "      <td>26</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs2067853</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>168</td>\n",
       "      <td>91</td>\n",
       "      <td>1407</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>1148</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>1407</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>1148</td>\n",
       "      <td>1136</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs2069832</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs2069833</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>188</td>\n",
       "      <td>104</td>\n",
       "      <td>1503</td>\n",
       "      <td>42</td>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>1211</td>\n",
       "      <td>...</td>\n",
       "      <td>104</td>\n",
       "      <td>1503</td>\n",
       "      <td>42</td>\n",
       "      <td>20</td>\n",
       "      <td>12</td>\n",
       "      <td>1211</td>\n",
       "      <td>1199</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs2069833</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs2069837</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>166</td>\n",
       "      <td>89</td>\n",
       "      <td>1323</td>\n",
       "      <td>23</td>\n",
       "      <td>170</td>\n",
       "      <td>12</td>\n",
       "      <td>1068</td>\n",
       "      <td>...</td>\n",
       "      <td>89</td>\n",
       "      <td>1323</td>\n",
       "      <td>23</td>\n",
       "      <td>170</td>\n",
       "      <td>12</td>\n",
       "      <td>1068</td>\n",
       "      <td>1056</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs2069845</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>68</td>\n",
       "      <td>34</td>\n",
       "      <td>480</td>\n",
       "      <td>13</td>\n",
       "      <td>43</td>\n",
       "      <td>6</td>\n",
       "      <td>378</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>480</td>\n",
       "      <td>13</td>\n",
       "      <td>43</td>\n",
       "      <td>6</td>\n",
       "      <td>378</td>\n",
       "      <td>372</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs2070744</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>219</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>178</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>219</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>178</td>\n",
       "      <td>178</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs3024490</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>72</td>\n",
       "      <td>38</td>\n",
       "      <td>653</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>653</td>\n",
       "      <td>13</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>543</td>\n",
       "      <td>542</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>rs34004222</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>188</td>\n",
       "      <td>104</td>\n",
       "      <td>1510</td>\n",
       "      <td>42</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>1218</td>\n",
       "      <td>...</td>\n",
       "      <td>104</td>\n",
       "      <td>1510</td>\n",
       "      <td>42</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>1218</td>\n",
       "      <td>1206</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs4846049</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>99</td>\n",
       "      <td>53</td>\n",
       "      <td>803</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>651</td>\n",
       "      <td>...</td>\n",
       "      <td>53</td>\n",
       "      <td>803</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>651</td>\n",
       "      <td>642</td>\n",
       "      <td>11</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>rs9651118</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>187</td>\n",
       "      <td>100</td>\n",
       "      <td>1505</td>\n",
       "      <td>24</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>1218</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>1505</td>\n",
       "      <td>24</td>\n",
       "      <td>184</td>\n",
       "      <td>12</td>\n",
       "      <td>1218</td>\n",
       "      <td>1206</td>\n",
       "      <td>16</td>\n",
       "      <td>28</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "REG          ID REF ALT  IH_ALT_CT_ACB  IH_ALT_CT_ASW  IH_ALT_CT_African  \\\n",
       "0     rs1071676   C   G            162             91               1298   \n",
       "1     rs1217401   A   G             45             25                317   \n",
       "2     rs1411040   C   T             27             13                264   \n",
       "3     rs1476413   C   T            154             88               1287   \n",
       "4     rs1518111   T   C             71             38                650   \n",
       "5     rs1554286   A   G             69             38                646   \n",
       "6     rs1554606   T   G             67             34                473   \n",
       "7     rs1800779   G   A             24             17                217   \n",
       "8     rs1800795   C   G             12              9                 22   \n",
       "9     rs1800796   G   C            171             97               1354   \n",
       "10    rs1800871   A   G             72             38                653   \n",
       "11    rs1800896   T   C            131             63               1016   \n",
       "12    rs1801131   T   G            153             84               1277   \n",
       "13    rs1801133   G   A            166             88               1384   \n",
       "14    rs1808593   G   T             43             26                430   \n",
       "15    rs1961495   C   T            154             93               1319   \n",
       "16    rs2043211   A   T            151             81               1232   \n",
       "17    rs2066992   G   T            176             97               1382   \n",
       "18    rs2067853   G   A            168             91               1407   \n",
       "19    rs2069832   A   G             12              9                 22   \n",
       "20    rs2069833   C   G            188            104               1503   \n",
       "21    rs2069833   C   T             12              8                 28   \n",
       "22    rs2069837   A   G            166             89               1323   \n",
       "23    rs2069845   G   A             68             34                480   \n",
       "24    rs2070744   C   T             24             17                219   \n",
       "25    rs3024490   A   C             72             38                653   \n",
       "26   rs34004222   G   A            188            104               1510   \n",
       "27    rs4846049   T   G             99             53                803   \n",
       "28    rs9651118   T   C            187            100               1505   \n",
       "\n",
       "REG  IH_ALT_CT_CA  IH_ALT_CT_EA  IH_ALT_CT_KS  IH_ALT_CT_Native African  ...  \\\n",
       "0              22           163             8                      1045  ...   \n",
       "1              17             5             4                       247  ...   \n",
       "2              11             7             1                       224  ...   \n",
       "3              27            18             8                      1045  ...   \n",
       "4               6             6             1                       541  ...   \n",
       "5              13            72             1                       539  ...   \n",
       "6              13             5             6                       372  ...   \n",
       "7               5             2             0                       176  ...   \n",
       "8               0             0             0                         1  ...   \n",
       "9              42            16            11                      1086  ...   \n",
       "10             13            72             1                       543  ...   \n",
       "11              6           124             7                       822  ...   \n",
       "12             20            18             8                      1040  ...   \n",
       "13             42           173            12                      1130  ...   \n",
       "14              3            60            10                       361  ...   \n",
       "15             23           156            12                      1072  ...   \n",
       "16             35            15            11                      1000  ...   \n",
       "17             24           145            12                      1109  ...   \n",
       "18             21            18            12                      1148  ...   \n",
       "19              0             0             0                         1  ...   \n",
       "20             42            20            12                      1211  ...   \n",
       "21              5             0             0                         8  ...   \n",
       "22             23           170            12                      1068  ...   \n",
       "23             13            43             6                       378  ...   \n",
       "24              4            29             0                       178  ...   \n",
       "25             13            72             1                       543  ...   \n",
       "26             42           184            12                      1218  ...   \n",
       "27             10             6             9                       651  ...   \n",
       "28             24           184            12                      1218  ...   \n",
       "\n",
       "REG  IH_REF_CT_ASW  IH_REF_CT_African  IH_REF_CT_CA  IH_REF_CT_EA  \\\n",
       "0               91               1298            22           163   \n",
       "1               25                317            17             5   \n",
       "2               13                264            11             7   \n",
       "3               88               1287            27            18   \n",
       "4               38                650             6             6   \n",
       "5               38                646            13            72   \n",
       "6               34                473            13             5   \n",
       "7               17                217             5             2   \n",
       "8                9                 22             0             0   \n",
       "9               97               1354            42            16   \n",
       "10              38                653            13            72   \n",
       "11              63               1016             6           124   \n",
       "12              84               1277            20            18   \n",
       "13              88               1384            42           173   \n",
       "14              26                430             3            60   \n",
       "15              93               1319            23           156   \n",
       "16              81               1232            35            15   \n",
       "17              97               1382            24           145   \n",
       "18              91               1407            21            18   \n",
       "19               9                 22             0             0   \n",
       "20             104               1503            42            20   \n",
       "21               8                 28             5             0   \n",
       "22              89               1323            23           170   \n",
       "23              34                480            13            43   \n",
       "24              17                219             4            29   \n",
       "25              38                653            13            72   \n",
       "26             104               1510            42           184   \n",
       "27              53                803            10             6   \n",
       "28             100               1505            24           184   \n",
       "\n",
       "REG  IH_REF_CT_KS  IH_REF_CT_Native African  \\\n",
       "0               8                      1045   \n",
       "1               4                       247   \n",
       "2               1                       224   \n",
       "3               8                      1045   \n",
       "4               1                       541   \n",
       "5               1                       539   \n",
       "6               6                       372   \n",
       "7               0                       176   \n",
       "8               0                         1   \n",
       "9              11                      1086   \n",
       "10              1                       543   \n",
       "11              7                       822   \n",
       "12              8                      1040   \n",
       "13             12                      1130   \n",
       "14             10                       361   \n",
       "15             12                      1072   \n",
       "16             11                      1000   \n",
       "17             12                      1109   \n",
       "18             12                      1148   \n",
       "19              0                         1   \n",
       "20             12                      1211   \n",
       "21              0                         8   \n",
       "22             12                      1068   \n",
       "23              6                       378   \n",
       "24              0                       178   \n",
       "25              1                       543   \n",
       "26             12                      1218   \n",
       "27              9                       651   \n",
       "28             12                      1218   \n",
       "\n",
       "REG  IH_REF_CT_Native African without KS  IH_REF_CT_SA  IH_REF_CT_SA and KS  \\\n",
       "0                                   1037            14                   22   \n",
       "1                                    243             3                    7   \n",
       "2                                    223             7                    8   \n",
       "3                                   1037            12                   20   \n",
       "4                                    540             5                    6   \n",
       "5                                    538             5                    6   \n",
       "6                                    366             3                    9   \n",
       "7                                    176             3                    3   \n",
       "8                                      1             0                    0   \n",
       "9                                   1075            14                   25   \n",
       "10                                   542             5                    6   \n",
       "11                                   815            11                   18   \n",
       "12                                  1032            12                   20   \n",
       "13                                  1118            16                   28   \n",
       "14                                   351             9                   19   \n",
       "15                                  1060            13                   25   \n",
       "16                                   989            14                   25   \n",
       "17                                  1097            14                   26   \n",
       "18                                  1136            16                   28   \n",
       "19                                     1             0                    0   \n",
       "20                                  1199            16                   28   \n",
       "21                                     8             0                    0   \n",
       "22                                  1056            13                   25   \n",
       "23                                   372             3                    9   \n",
       "24                                   178             3                    3   \n",
       "25                                   542             5                    6   \n",
       "26                                  1206            16                   28   \n",
       "27                                   642            11                   20   \n",
       "28                                  1206            16                   28   \n",
       "\n",
       "REG  IH_REF_CT_WA  \n",
       "0             184  \n",
       "1              39  \n",
       "2              21  \n",
       "3             176  \n",
       "4             108  \n",
       "5              22  \n",
       "6              12  \n",
       "7              28  \n",
       "8               1  \n",
       "9             153  \n",
       "10             22  \n",
       "11            120  \n",
       "12            204  \n",
       "13            210  \n",
       "14             65  \n",
       "15            217  \n",
       "16             30  \n",
       "17            198  \n",
       "18             40  \n",
       "19              0  \n",
       "20             42  \n",
       "21              0  \n",
       "22            145  \n",
       "23             67  \n",
       "24             17  \n",
       "25            109  \n",
       "26            234  \n",
       "27             23  \n",
       "28            232  \n",
       "\n",
       "[29 rows x 25 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Format in-house allele count data in format suitable for Fisher's test\n",
    "\n",
    "ih_fishers_data = collated_count_data[collated_count_data.ID != \".\"] # Drop ambiguous variant IDs\n",
    "ih_fishers_data = ih_fishers_data.drop_duplicates([\"ID\",\"REF\",\"ALT\",\"REG\"]) # Drop duplicate entries\n",
    "ih_fishers_data = ih_fishers_data.pivot(index=[\"ID\",\"REF\",\"ALT\"], columns=\"REG\", values=[\"ALT_CTS\",\"REF_CTS\"]) # Pivot data\n",
    "\n",
    "# Separate alternate and reference count data into different dataframes to facilate renaming of count columns appropriately\n",
    "ih_fishers_data_alt = ih_fishers_data[['ALT_CTS']].droplevel(level=0, axis=1).reset_index() \n",
    "fishers_data_ref = ih_fishers_data[['REF_CTS']].droplevel(level=0, axis=1).reset_index()\n",
    "\n",
    "ih_fishers_data_alt = add_prefix_dataframe_col_names(fishers_data_ref, inhouse_populations,\"IH_ALT_CT_\")\n",
    "ih_fishers_data_ref = add_prefix_dataframe_col_names(fishers_data_ref, inhouse_populations,\"IH_REF_CT_\")\n",
    "\n",
    "# Merge renamed alternate and reference count data\n",
    "ih_fishers_data_renamed = ih_fishers_data_alt.merge(ih_fishers_data_ref, on=[\"ID\",\"REF\",\"ALT\"])\n",
    "ih_fishers_data_renamed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     C\n",
       "1     A\n",
       "2     C\n",
       "3     C\n",
       "4     T\n",
       "5     A\n",
       "6     T\n",
       "7     G\n",
       "8     C\n",
       "9     G\n",
       "10    A\n",
       "11    T\n",
       "12    T\n",
       "13    G\n",
       "14    G\n",
       "15    C\n",
       "16    A\n",
       "17    G\n",
       "18    G\n",
       "19    A\n",
       "20    C\n",
       "21    C\n",
       "22    A\n",
       "23    G\n",
       "24    C\n",
       "25    A\n",
       "26    G\n",
       "27    T\n",
       "28    T\n",
       "Name: REF, dtype: object"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ih_fishers_data_renamed['REF']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REG          ID REF ALT  ALFA_REF_CT_African  ALFA_REF_CT_African American  \\\n",
      "0     rs1071676   C   G               2489.0                        2392.0   \n",
      "1     rs1217401   A   G               3779.0                        3697.0   \n",
      "2     rs1411040   C   T               1218.0                        1169.0   \n",
      "3     rs1476413   C   T               7510.0                        7223.0   \n",
      "4     rs1518111   T   C               3285.0                        3154.0   \n",
      "5     rs1554286   A   G               3188.0                        3069.0   \n",
      "6     rs1554606   T   G               1716.0                        1651.0   \n",
      "7     rs1800779   G   A                581.0                         568.0   \n",
      "8     rs1800795   C   G                195.0                         194.0   \n",
      "9     rs1800796   G   C               2657.0                        2552.0   \n",
      "10    rs1800871   A   G               3248.0                        3117.0   \n",
      "11    rs1800896   T   C               5586.0                        5349.0   \n",
      "12    rs1801131   T   G               9531.0                        9212.0   \n",
      "13    rs1801133   G   A              10304.0                        9924.0   \n",
      "14    rs1808593   G   T               1648.0                        1606.0   \n",
      "15    rs1961495   C   T               4458.0                        4308.0   \n",
      "16    rs2043211   A   T               3446.0                        3312.0   \n",
      "17    rs2066992   G   T               7491.0                        7236.0   \n",
      "18    rs2067853   G   A               6158.0                        5936.0   \n",
      "19    rs2069832   A   G                432.0                         431.0   \n",
      "20    rs2069833   C   G                 71.0                          70.0   \n",
      "21    rs2069833   C   T                 71.0                          70.0   \n",
      "22    rs2069837   A   G               6546.0                        6306.0   \n",
      "23    rs2069845   G   A               2228.0                        2158.0   \n",
      "24    rs2070744   C   T                453.0                         439.0   \n",
      "25    rs3024490   A   C               1439.0                        1378.0   \n",
      "26   rs34004222   G   A               6962.0                        6720.0   \n",
      "27    rs4846049   T   G               3385.0                        3272.0   \n",
      "28    rs9651118   T   C               9857.0                        9504.0   \n",
      "\n",
      "REG  ALFA_REF_CT_African Others  ALFA_REF_CT_Asian  ALFA_REF_CT_East Asian  \\\n",
      "0                          97.0              106.0                    83.0   \n",
      "1                          82.0             6623.0                  4735.0   \n",
      "2                          49.0               10.0                     7.0   \n",
      "3                         287.0             3080.0                  2562.0   \n",
      "4                         131.0             4270.0                  3161.0   \n",
      "5                         119.0              208.0                   137.0   \n",
      "6                          65.0                9.0                     5.0   \n",
      "7                          13.0               13.0                    10.0   \n",
      "8                           1.0                0.0                     0.0   \n",
      "9                         105.0               37.0                    28.0   \n",
      "10                        131.0             4662.0                  3469.0   \n",
      "11                        237.0             3690.0                  3001.0   \n",
      "12                        319.0              666.0                   531.0   \n",
      "13                        380.0             2635.0                  1954.0   \n",
      "14                         42.0              138.0                   106.0   \n",
      "15                        150.0              143.0                   114.0   \n",
      "16                        134.0             1753.0                  1427.0   \n",
      "17                        255.0              211.0                   144.0   \n",
      "18                        222.0              576.0                   466.0   \n",
      "19                          1.0                5.0                     2.0   \n",
      "20                          1.0                0.0                     0.0   \n",
      "21                          1.0                0.0                     0.0   \n",
      "22                        240.0              570.0                   449.0   \n",
      "23                         70.0                1.0                     1.0   \n",
      "24                         14.0               15.0                    12.0   \n",
      "25                         61.0              140.0                    97.0   \n",
      "26                        242.0             6680.0                  4790.0   \n",
      "27                        113.0               90.0                    69.0   \n",
      "28                        353.0              457.0                   365.0   \n",
      "\n",
      "REG  ALFA_REF_CT_European  ALFA_REF_CT_Latin American 1  ...  \\\n",
      "0                 13426.0                         120.0  ...   \n",
      "1                221196.0                         949.0  ...   \n",
      "2                  3680.0                          43.0  ...   \n",
      "3                201818.0                         886.0  ...   \n",
      "4                 33302.0                         123.0  ...   \n",
      "5                 11916.0                          93.0  ...   \n",
      "6                 71021.0                         282.0  ...   \n",
      "7                  9780.0                          58.0  ...   \n",
      "8                  6286.0                          25.0  ...   \n",
      "9                 16700.0                         128.0  ...   \n",
      "10                59475.0                         307.0  ...   \n",
      "11               140244.0                         727.0  ...   \n",
      "12               145439.0                         998.0  ...   \n",
      "13               209760.0                        1056.0  ...   \n",
      "14                40575.0                         198.0  ...   \n",
      "15                36637.0                         490.0  ...   \n",
      "16                66215.0                         558.0  ...   \n",
      "17               162183.0                         870.0  ...   \n",
      "18                84208.0                         522.0  ...   \n",
      "19                49002.0                         209.0  ...   \n",
      "20                 4645.0                           6.0  ...   \n",
      "21                 4645.0                           6.0  ...   \n",
      "22               163294.0                         892.0  ...   \n",
      "23                 6588.0                          37.0  ...   \n",
      "24                11148.0                          51.0  ...   \n",
      "25                12393.0                          82.0  ...   \n",
      "26               188175.0                        1172.0  ...   \n",
      "27                10050.0                         169.0  ...   \n",
      "28               149352.0                         826.0  ...   \n",
      "\n",
      "REG  ALFA_ALT_CT_African Others  ALFA_ALT_CT_Asian  ALFA_ALT_CT_East Asian  \\\n",
      "0                          17.0                6.0                     3.0   \n",
      "1                         340.0              403.0                   291.0   \n",
      "2                         195.0              162.0                   105.0   \n",
      "3                          39.0              894.0                   640.0   \n",
      "4                         155.0             2014.0                  1323.0   \n",
      "5                         137.0               86.0                    53.0   \n",
      "6                         103.0              585.0                   467.0   \n",
      "7                         121.0              117.0                    94.0   \n",
      "8                         113.0              112.0                    86.0   \n",
      "9                           9.0               75.0                    58.0   \n",
      "10                        161.0             2172.0                  1443.0   \n",
      "11                         81.0              248.0                   193.0   \n",
      "12                         49.0              218.0                   145.0   \n",
      "13                         30.0             1349.0                  1230.0   \n",
      "14                        198.0              562.0                   462.0   \n",
      "15                         32.0               57.0                    50.0   \n",
      "16                         40.0             1517.0                  1193.0   \n",
      "17                         21.0              537.0                   426.0   \n",
      "18                          8.0               66.0                    52.0   \n",
      "19                        217.0              625.0                   496.0   \n",
      "20                          0.0                0.0                     0.0   \n",
      "21                         29.0                4.0                     2.0   \n",
      "22                         20.0              124.0                    83.0   \n",
      "23                        154.0              115.0                    87.0   \n",
      "24                        100.0              139.0                   116.0   \n",
      "25                         35.0               36.0                    21.0   \n",
      "26                          0.0                0.0                     0.0   \n",
      "27                         77.0              286.0                   259.0   \n",
      "28                          3.0              327.0                   243.0   \n",
      "\n",
      "REG  ALFA_ALT_CT_European  ALFA_ALT_CT_Latin American 1  \\\n",
      "0                  4168.0                          26.0   \n",
      "1                105092.0                         685.0   \n",
      "2                 19802.0                         269.0   \n",
      "3                 73888.0                         248.0   \n",
      "4                121504.0                         341.0   \n",
      "5                 53284.0                         307.0   \n",
      "6                 92771.0                         426.0   \n",
      "7                 16382.0                         112.0   \n",
      "8                  8000.0                         121.0   \n",
      "9                   882.0                          18.0   \n",
      "10               190523.0                         721.0   \n",
      "11               125934.0                         403.0   \n",
      "12                66341.0                         280.0   \n",
      "13               112566.0                         432.0   \n",
      "14               119161.0                         540.0   \n",
      "15                 4591.0                          72.0   \n",
      "16                31651.0                         232.0   \n",
      "17                 9301.0                         116.0   \n",
      "18                39304.0                         156.0   \n",
      "19                73610.0                         543.0   \n",
      "20                    0.0                           0.0   \n",
      "21                 3831.0                           0.0   \n",
      "22                13656.0                          72.0   \n",
      "23                 7706.0                         117.0   \n",
      "24                18706.0                          95.0   \n",
      "25                38477.0                         176.0   \n",
      "26                  761.0                           0.0   \n",
      "27                20028.0                         273.0   \n",
      "28                38926.0                         146.0   \n",
      "\n",
      "REG ALFA_ALT_CT_Latin American 2 ALFA_ALT_CT_Other ALFA_ALT_CT_Other Asian  \\\n",
      "0                           87.0             144.0                     3.0   \n",
      "1                         1809.0            7133.0                   112.0   \n",
      "2                         2266.0            1668.0                    57.0   \n",
      "3                         1150.0            2865.0                   254.0   \n",
      "4                          668.0            6964.0                   691.0   \n",
      "5                         2256.0            3341.0                    33.0   \n",
      "6                         4817.0            3910.0                   118.0   \n",
      "7                          555.0            1093.0                    23.0   \n",
      "8                          500.0             503.0                    26.0   \n",
      "9                          161.0             113.0                    17.0   \n",
      "10                        4118.0           10367.0                   729.0   \n",
      "11                        2045.0            3545.0                    55.0   \n",
      "12                         964.0            4136.0                    73.0   \n",
      "13                        3277.0            6838.0                   119.0   \n",
      "14                        5222.0            5651.0                   100.0   \n",
      "15                        1139.0             598.0                     7.0   \n",
      "16                         272.0            2655.0                   324.0   \n",
      "17                        2560.0             912.0                   111.0   \n",
      "18                         935.0            1678.0                    14.0   \n",
      "19                        5372.0            4244.0                   129.0   \n",
      "20                           0.0               0.0                     0.0   \n",
      "21                           0.0             143.0                     2.0   \n",
      "22                         478.0             601.0                    41.0   \n",
      "23                         449.0             628.0                    28.0   \n",
      "24                         453.0             636.0                    23.0   \n",
      "25                        1737.0             899.0                    15.0   \n",
      "26                           2.0              53.0                     0.0   \n",
      "27                        3473.0            1979.0                    27.0   \n",
      "28                        2350.0            1582.0                    84.0   \n",
      "\n",
      "REG  ALFA_ALT_CT_South Asian  ALFA_ALT_CT_Total  \n",
      "0                       13.0             4901.0  \n",
      "1                     1715.0           125456.0  \n",
      "2                       79.0            30452.0  \n",
      "3                     2152.0            82675.0  \n",
      "4                      169.0           136517.0  \n",
      "5                     3257.0            67359.0  \n",
      "6                     3728.0           109085.0  \n",
      "7                       88.0            21514.0  \n",
      "8                       84.0            12071.0  \n",
      "9                       39.0             1577.0  \n",
      "10                     214.0           212875.0  \n",
      "11                    1517.0           136458.0  \n",
      "12                    2076.0            75950.0  \n",
      "13                     875.0           126753.0  \n",
      "14                     156.0           136356.0  \n",
      "15                      37.0             7284.0  \n",
      "16                      95.0            37350.0  \n",
      "17                    1474.0            15581.0  \n",
      "18                      30.0            42603.0  \n",
      "19                     150.0            90274.0  \n",
      "20                       0.0                0.0  \n",
      "21                       4.0             4573.0  \n",
      "22                     792.0            16273.0  \n",
      "23                      80.0            13565.0  \n",
      "24                      74.0            22596.0  \n",
      "25                    2923.0            45635.0  \n",
      "26                       0.0              824.0  \n",
      "27                      42.0            28646.0  \n",
      "28                    1321.0            45103.0  \n",
      "\n",
      "[29 rows x 30 columns]\n"
     ]
    }
   ],
   "source": [
    "# Extract ALFA data alternate allele information that corresponds with that of the in-house alternate allele count data\n",
    "\n",
    "# Iterate through in-house variant data and extract variant ID, alternate allele and regional information:\n",
    "corresponding_ALFA_count_df = pd.DataFrame()\n",
    "for index, row in ih_fishers_data_renamed.drop_duplicates(subset=[\"ID\", \"REF\", \"ALT\"]).iterrows():\n",
    "    variant_ID = row[\"ID\"]\n",
    "    variant_ref = row[\"REF\"]\n",
    "    variant_alt = row[\"ALT\"]\n",
    "    \n",
    "    # Extract corresponding alternate allele information from ALFA count dataframe:\n",
    "    corresponding_ALFA_pop_count_df = pd.DataFrame()\n",
    "    ALFA_populations = []\n",
    "    for population in set(normalised_variant_count_data.Population.values): # Loop through every unique population in the ALFA count dataframe\n",
    "        ALFA_populations.append(population)\n",
    "        corresponding_ALFA_pop_count = normalised_variant_count_data[(normalised_variant_count_data.variant_id == variant_ID) & (normalised_variant_count_data.reference_allele == variant_ref) & (normalised_variant_count_data.Population == population)] # Extract count information for a particular variant in a particular population\n",
    "        corresponding_ALFA_pop_count_df[\"ID\"] = corresponding_ALFA_pop_count.variant_id.values\n",
    "        corresponding_ALFA_pop_count_df[\"REG\"] = corresponding_ALFA_pop_count.Population.values\n",
    "        corresponding_ALFA_pop_count_df[\"REF\"] = corresponding_ALFA_pop_count.reference_allele.values\n",
    "        corresponding_ALFA_pop_count_df[\"ALT\"] = variant_alt\n",
    "        corresponding_ALFA_pop_count_df[\"ALFA_REF_CTS\"] = corresponding_ALFA_pop_count[variant_ref].values\n",
    "        corresponding_ALFA_pop_count_df[\"ALFA_ALT_CTS\"] = corresponding_ALFA_pop_count[variant_alt].values\n",
    "        \n",
    "        corresponding_ALFA_count_df = pd.concat([corresponding_ALFA_count_df, corresponding_ALFA_pop_count_df]).reset_index(drop=True)\n",
    "\n",
    "corresponding_ALFA_count_pivot = corresponding_ALFA_count_df.pivot(index=[\"ID\",\"REF\",\"ALT\"], columns=\"REG\", values=[\"ALFA_REF_CTS\",\"ALFA_ALT_CTS\"]) # Pivot data\n",
    "\n",
    "corresponding_ALFA_count_alt = corresponding_ALFA_count_pivot[['ALFA_ALT_CTS']].droplevel(level=0, axis=1).reset_index() \n",
    "corresponding_ALFA_count_ref = corresponding_ALFA_count_pivot[['ALFA_REF_CTS']].droplevel(level=0, axis=1).reset_index()\n",
    "        \n",
    "ALFA_data_alt = add_prefix_dataframe_col_names(corresponding_ALFA_count_alt, ALFA_populations,\"ALFA_ALT_CT_\")\n",
    "ALFA_data_ref = add_prefix_dataframe_col_names(corresponding_ALFA_count_ref, ALFA_populations,\"ALFA_REF_CT_\")\n",
    "\n",
    "ALFA_data = pd.concat([ALFA_data_ref, ALFA_data_alt], axis=1)\n",
    "print (ALFA_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "23b4a3e8622309bcc6db3d5cc6eb73d60ab98d9ec23bad6a26b709981ccb403a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
