{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform Fisher's Exact Tests to test if differences in allele frequencies for population groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules and packages\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from scipy.stats import fisher_exact\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from warnings import simplefilter\n",
    "from statsmodels.stats.multitest import multipletests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set constants and functions \n",
    "\n",
    "home_path = str(os.path.dirname(os.getcwd()))\n",
    "population_clusters = [\"SUPER\",\"SUB\"]\n",
    "\n",
    "gene_location_df = pd.read_csv(os.path.join(home_path, \"Data_descriptions\", \"locations.csv\"))\n",
    "genes = gene_location_df.location_name\n",
    "\n",
    "sample_info_df = pd.read_csv(os.path.join(home_path, \"Data_descriptions\", \"samples.csv\"))\n",
    "sub_populations = sample_info_df.SUB.unique().tolist()\n",
    "regional_classification = {\"ACB\":\"ACB\",\"ASW\":\"ASW\",\"GWD\":\"WA\",\"ESN\":\"WA\",\"MSL\":\"WA\",\"MbutiPygmy\":\"CA\",\"BiakaPygmy\":\"CA\",\"Mandenka\":\"WA\",\"Yoruba\":\"WA\",\"San\":\"KS\",\"BantuSouthAfrica\":\"SA\",\"BantuKenya\":\"EA\",\"YRI\":\"WA\",\"LWK\":\"EA\"} # SA = South Africa, CA = Central Africa, WA = West Africa, EA = East Africa, ACB = African Carribean in Barbados, ASW = African American, KS = Khoi-San\n",
    "\n",
    "def add_prefix_dataframe_col_names(dataframe, columns_list, prefix):\n",
    "    rename_dict = dict()\n",
    "    for line in columns_list:\n",
    "        key = line\n",
    "        if key not in rename_dict:\n",
    "            rename_dict[key] = str()\n",
    "        rename_dict[key] = prefix + \"{}\".format(line)\n",
    "    dataframe = dataframe.rename(columns = rename_dict)\n",
    "    return dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_CTS</th>\n",
       "      <th>REF_CTS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SUB_POP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>rs1185300901</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>0</td>\n",
       "      <td>188</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>rs552586867</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>188</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>rs59409892</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>19</td>\n",
       "      <td>169</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>rs535182970</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>rs56406633</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>ACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>21</td>\n",
       "      <td>rs981058453</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>21</td>\n",
       "      <td>rs182058038</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>21</td>\n",
       "      <td>rs7278343</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>21</td>\n",
       "      <td>83</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>21</td>\n",
       "      <td>rs151281307</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>21</td>\n",
       "      <td>rs554778511</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>OLIG2</td>\n",
       "      <td>ASW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326830 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM            ID REF ALT  ALT_CTS  REF_CTS    GENE SUB_POP\n",
       "0        13  rs1185300901   C  CT        0      188  COL4A1     ACB\n",
       "1        13   rs552586867   C   G        0      188  COL4A1     ACB\n",
       "2        13    rs59409892   C   G       19      169  COL4A1     ACB\n",
       "3        13   rs535182970   G   C        1      187  COL4A1     ACB\n",
       "4        13    rs56406633   A   G        1      187  COL4A1     ACB\n",
       "..      ...           ...  ..  ..      ...      ...     ...     ...\n",
       "158      21   rs981058453   C   T        0      104   OLIG2     ASW\n",
       "159      21   rs182058038   T   A        1      103   OLIG2     ASW\n",
       "160      21     rs7278343   A   G       21       83   OLIG2     ASW\n",
       "161      21   rs151281307  AT   A        1      103   OLIG2     ASW\n",
       "162      21   rs554778511   G   A        0      104   OLIG2     ASW\n",
       "\n",
       "[326830 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import allele count data\n",
    "\n",
    "collated_count_data = pd.DataFrame()\n",
    "for gene in genes:\n",
    "    for sub_population in sub_populations:\n",
    "        allele_count_path = os.path.join(home_path,\"Data\", \"SUB\", \"ALL_{}.{}.acount\".format(gene, sub_population))\n",
    "        if os.path.exists(allele_count_path):\n",
    "            allele_count_df = pd.read_csv(allele_count_path, sep=\"\\t\")\n",
    "            allele_count_df[\"REF_CTS\"] = allele_count_df[\"OBS_CT\"] - allele_count_df[\"ALT_CTS\"]\n",
    "            allele_count_df[\"GENE\"] = gene\n",
    "            allele_count_df[\"SUB_POP\"] = sub_population\n",
    "        collated_count_data = pd.concat([collated_count_data, allele_count_df]).drop(columns=\"OBS_CT\")\n",
    "\n",
    "collated_count_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format allele count data in format suitable for Fisher's test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SUB_POP  #CHROM            ID REF ALT  ALT_CT_ACB  ALT_CT_ASW  \\\n",
      "0             1  rs1000680496   G   A           0           0   \n",
      "1             1  rs1001911383   C   T           0           0   \n",
      "2             1  rs1002151986   A   C           0           0   \n",
      "3             1  rs1002155376   G   T           0           0   \n",
      "4             1  rs1002228547   G   C           1           0   \n",
      "...         ...           ...  ..  ..         ...         ...   \n",
      "20302        21   rs972037300   C   G           0           0   \n",
      "20303        21   rs975220347   C   T           0           0   \n",
      "20304        21   rs981058453   C   T           0           0   \n",
      "20305        21   rs997790265   C  CG           0           0   \n",
      "20306        21     rs9982080   G   T          29          16   \n",
      "\n",
      "SUB_POP  ALT_CT_BantuKenya  ALT_CT_BantuSouthAfrica  ALT_CT_BiakaPygmy  \\\n",
      "0                        0                        0                  0   \n",
      "1                        0                        0                  0   \n",
      "2                        0                        0                  0   \n",
      "3                        0                        0                  0   \n",
      "4                        0                        0                  0   \n",
      "...                    ...                      ...                ...   \n",
      "20302                    0                        0                  0   \n",
      "20303                    0                        0                  0   \n",
      "20304                    0                        0                  0   \n",
      "20305                    0                        0                  0   \n",
      "20306                    3                        4                  5   \n",
      "\n",
      "SUB_POP  ALT_CT_ESN  ...  REF_CT_BiakaPygmy  REF_CT_ESN  REF_CT_GWD  \\\n",
      "0                 0  ...                 42         206         231   \n",
      "1                 0  ...                 36         204         228   \n",
      "2                 0  ...                 42         206         232   \n",
      "3                 0  ...                 42         206         232   \n",
      "4                 0  ...                 42         206         232   \n",
      "...             ...  ...                ...         ...         ...   \n",
      "20302             0  ...                 42         206         232   \n",
      "20303             0  ...                 42         206         232   \n",
      "20304             0  ...                 42         206         232   \n",
      "20305             0  ...                 42         204         232   \n",
      "20306            23  ...                 37         183         202   \n",
      "\n",
      "SUB_POP  REF_CT_LWK  REF_CT_MSL  REF_CT_Mandenka  REF_CT_MbutiPygmy  \\\n",
      "0               184         166               40                 24   \n",
      "1               182         166               38                 22   \n",
      "2               184         166               40                 24   \n",
      "3               184         166               40                 24   \n",
      "4               184         166               40                 24   \n",
      "...             ...         ...              ...                ...   \n",
      "20302           184         166               40                 24   \n",
      "20303           184         166               40                 24   \n",
      "20304           184         166               40                 24   \n",
      "20305           182         166               38                 24   \n",
      "20306           138         146               39                 24   \n",
      "\n",
      "SUB_POP  REF_CT_San  REF_CT_YRI  REF_CT_Yoruba  \n",
      "0                12         234             42  \n",
      "1                12         234             36  \n",
      "2                12         234             42  \n",
      "3                12         234             42  \n",
      "4                12         234             42  \n",
      "...             ...         ...            ...  \n",
      "20302            12         234             42  \n",
      "20303            12         234             42  \n",
      "20304            12         234             41  \n",
      "20305            12         234             42  \n",
      "20306            10         200             34  \n",
      "\n",
      "[20307 rows x 32 columns]\n"
     ]
    }
   ],
   "source": [
    "fishers_data = collated_count_data[collated_count_data.ID != \".\"] # Drop ambiguous variant IDs\n",
    "fishers_data = fishers_data.drop_duplicates([\"#CHROM\",\"ID\",\"REF\",\"ALT\",\"SUB_POP\"]) # Drop duplicate entries\n",
    "fishers_data = fishers_data.pivot(index=[\"#CHROM\",\"ID\",\"REF\",\"ALT\"], columns=\"SUB_POP\", values=[\"ALT_CTS\",\"REF_CTS\"]) # Pivot data\n",
    "\n",
    "# Separate alternate and reference count data into different dataframes to facilate renaming of count columns appropriately\n",
    "fishers_data_alt = fishers_data[['ALT_CTS']].droplevel(level=0, axis=1).reset_index() \n",
    "fishers_data_ref = fishers_data[['REF_CTS']].droplevel(level=0, axis=1).reset_index()\n",
    "\n",
    "fishers_data_alt = add_prefix_dataframe_col_names(fishers_data_alt, sub_populations,\"ALT_CT_\")\n",
    "fishers_data_ref = add_prefix_dataframe_col_names(fishers_data_ref, sub_populations,\"REF_CT_\")\n",
    "\n",
    "# Merge renamed alternate and reference count data\n",
    "fishers_data_renamed = fishers_data_alt.merge(fishers_data_ref, on=[\"#CHROM\",\"ID\",\"REF\",\"ALT\"])\n",
    "print(fishers_data_renamed)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform Fisher's Exact test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ACB', 'GWD'),\n",
       " ('ACB', 'ESN'),\n",
       " ('ACB', 'MSL'),\n",
       " ('ACB', 'MbutiPygmy'),\n",
       " ('ACB', 'BiakaPygmy'),\n",
       " ('ACB', 'Mandenka'),\n",
       " ('ACB', 'Yoruba'),\n",
       " ('ACB', 'San'),\n",
       " ('ACB', 'BantuSouthAfrica'),\n",
       " ('ACB', 'BantuKenya'),\n",
       " ('ACB', 'YRI'),\n",
       " ('ACB', 'LWK'),\n",
       " ('ACB', 'ASW'),\n",
       " ('GWD', 'ESN'),\n",
       " ('GWD', 'MSL'),\n",
       " ('GWD', 'MbutiPygmy'),\n",
       " ('GWD', 'BiakaPygmy'),\n",
       " ('GWD', 'Mandenka'),\n",
       " ('GWD', 'Yoruba'),\n",
       " ('GWD', 'San'),\n",
       " ('GWD', 'BantuSouthAfrica'),\n",
       " ('GWD', 'BantuKenya'),\n",
       " ('GWD', 'YRI'),\n",
       " ('GWD', 'LWK'),\n",
       " ('GWD', 'ASW'),\n",
       " ('ESN', 'MSL'),\n",
       " ('ESN', 'MbutiPygmy'),\n",
       " ('ESN', 'BiakaPygmy'),\n",
       " ('ESN', 'Mandenka'),\n",
       " ('ESN', 'Yoruba'),\n",
       " ('ESN', 'San'),\n",
       " ('ESN', 'BantuSouthAfrica'),\n",
       " ('ESN', 'BantuKenya'),\n",
       " ('ESN', 'YRI'),\n",
       " ('ESN', 'LWK'),\n",
       " ('ESN', 'ASW'),\n",
       " ('MSL', 'MbutiPygmy'),\n",
       " ('MSL', 'BiakaPygmy'),\n",
       " ('MSL', 'Mandenka'),\n",
       " ('MSL', 'Yoruba'),\n",
       " ('MSL', 'San'),\n",
       " ('MSL', 'BantuSouthAfrica'),\n",
       " ('MSL', 'BantuKenya'),\n",
       " ('MSL', 'YRI'),\n",
       " ('MSL', 'LWK'),\n",
       " ('MSL', 'ASW'),\n",
       " ('MbutiPygmy', 'BiakaPygmy'),\n",
       " ('MbutiPygmy', 'Mandenka'),\n",
       " ('MbutiPygmy', 'Yoruba'),\n",
       " ('MbutiPygmy', 'San'),\n",
       " ('MbutiPygmy', 'BantuSouthAfrica'),\n",
       " ('MbutiPygmy', 'BantuKenya'),\n",
       " ('MbutiPygmy', 'YRI'),\n",
       " ('MbutiPygmy', 'LWK'),\n",
       " ('MbutiPygmy', 'ASW'),\n",
       " ('BiakaPygmy', 'Mandenka'),\n",
       " ('BiakaPygmy', 'Yoruba'),\n",
       " ('BiakaPygmy', 'San'),\n",
       " ('BiakaPygmy', 'BantuSouthAfrica'),\n",
       " ('BiakaPygmy', 'BantuKenya'),\n",
       " ('BiakaPygmy', 'YRI'),\n",
       " ('BiakaPygmy', 'LWK'),\n",
       " ('BiakaPygmy', 'ASW'),\n",
       " ('Mandenka', 'Yoruba'),\n",
       " ('Mandenka', 'San'),\n",
       " ('Mandenka', 'BantuSouthAfrica'),\n",
       " ('Mandenka', 'BantuKenya'),\n",
       " ('Mandenka', 'YRI'),\n",
       " ('Mandenka', 'LWK'),\n",
       " ('Mandenka', 'ASW'),\n",
       " ('Yoruba', 'San'),\n",
       " ('Yoruba', 'BantuSouthAfrica'),\n",
       " ('Yoruba', 'BantuKenya'),\n",
       " ('Yoruba', 'YRI'),\n",
       " ('Yoruba', 'LWK'),\n",
       " ('Yoruba', 'ASW'),\n",
       " ('San', 'BantuSouthAfrica'),\n",
       " ('San', 'BantuKenya'),\n",
       " ('San', 'YRI'),\n",
       " ('San', 'LWK'),\n",
       " ('San', 'ASW'),\n",
       " ('BantuSouthAfrica', 'BantuKenya'),\n",
       " ('BantuSouthAfrica', 'YRI'),\n",
       " ('BantuSouthAfrica', 'LWK'),\n",
       " ('BantuSouthAfrica', 'ASW'),\n",
       " ('BantuKenya', 'YRI'),\n",
       " ('BantuKenya', 'LWK'),\n",
       " ('BantuKenya', 'ASW'),\n",
       " ('YRI', 'LWK'),\n",
       " ('YRI', 'ASW'),\n",
       " ('LWK', 'ASW')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate all possible combinations of populations for comparison\n",
    "combinations = list(itertools.combinations(sub_populations,2))\n",
    "combinations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nan 1.0\n"
     ]
    }
   ],
   "source": [
    "oddsratio, pvalue = fisher_exact([[186,228],[0,0]])\n",
    "print (oddsratio, pvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate fisher's exact odds ratios and p-values for each variant for the different population combinations\n",
    "\n",
    "simplefilter(action=\"ignore\", category=pd.errors.PerformanceWarning)\n",
    "\n",
    "fishers_results = pd.DataFrame()\n",
    "for index, row in fishers_data_renamed.iterrows():\n",
    "    fisher_variant_combination_results = pd.DataFrame()\n",
    "    for combination in combinations:\n",
    "        first_pop = combination[0]\n",
    "        second_pop = combination[1]\n",
    "        alt_1 = row[\"ALT_CT_{}\".format(first_pop)]\n",
    "        ref_1 = row[\"REF_CT_{}\".format(first_pop)]\n",
    "        alt_2 = row[\"ALT_CT_{}\".format(second_pop)]\n",
    "        ref_2 = row[\"REF_CT_{}\".format(second_pop)]\n",
    "        oddsratio, pvalue = fisher_exact([[alt_1,alt_2],[ref_1,ref_2]])\n",
    "        fisher_variant_combination_results[\"CHROM\"] = [row[\"#CHROM\"]]\n",
    "        fisher_variant_combination_results[\"ID\"] = row[\"ID\"]\n",
    "        fisher_variant_combination_results[\"REF\"] = row[\"REF\"]\n",
    "        fisher_variant_combination_results[\"ALT\"] = row[\"ALT\"]\n",
    "        fisher_variant_combination_results[\"PVALUE_{}_{}\".format(first_pop, second_pop)] = pvalue\n",
    "        fisher_variant_combination_results[\"OR_{}_{}\".format(first_pop, second_pop)] = oddsratio\n",
    "    fishers_results = pd.concat([fishers_results,fisher_variant_combination_results])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Fisher's test results to CSV file\n",
    "\n",
    "fishers_results.reset_index(drop=True).to_csv(os.path.join(home_path, \"Data\", \"FishersExactTest\", \"Fishers_results_subpopulations.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct for multiple testing by controlling false discovery rate (FDR) using Benjamini and Hochberg\n",
    "\n",
    "# Generate a list with column names for all subpopulation p-values in the fishers_results dataframe\n",
    "p_value_combinations_list = []\n",
    "for combination in combinations:\n",
    "    first_pop = combination[0]\n",
    "    second_pop = combination[1]\n",
    "    p_value_combinations = \"PVALUE_{}_{}\".format(first_pop, second_pop)\n",
    "    p_value_combinations_list.append(p_value_combinations)\n",
    "\n",
    "# Method 1: Segregate the data by subpopulation and correct for multiple testing within a subpopulation\n",
    "\n",
    "corrected_pvalues = pd.DataFrame()\n",
    "for item in p_value_combinations_list:\n",
    "    multipletests_input = fishers_results[item]\n",
    "    multipletests_results = multipletests(multipletests_input, alpha=0.05, method=\"fdr_bh\")[1]\n",
    "    corrected_pvalues[\"{}_FDR\".format(item)] = multipletests_results\n",
    "\n",
    "corrected_pvalues = pd.concat([fishers_results.reset_index(drop=True), corrected_pvalues.reset_index(drop=True)], axis=1)\n",
    "corrected_pvalues.to_csv(os.path.join(home_path, \"Data\", \"FishersExactTest\", \"Fishers_multipletestcorrection_fdrbh_pd.csv\"))\n",
    "\n",
    "# Method 2: Correct for multiple testing for the entire dataset without segregating the data per subpopulation\n",
    "\n",
    "fishers_results_melt = pd.melt(fishers_results, id_vars=[\"CHROM\",\"ID\",\"REF\",\"ALT\"], value_vars=p_value_combinations_list)\n",
    "multipletests_input = fishers_results_melt[[\"value\"]].values.flatten().tolist()\n",
    "multipletests_results = multipletests(multipletests_input, alpha=0.05, method=\"fdr_bh\")\n",
    "multipletests_results_df = pd.DataFrame(multipletests_results[1])\n",
    "multipletests_results_info = pd.concat([fishers_results_melt.reset_index(drop=True), multipletests_results_df.reset_index(drop=True)], axis=1)\n",
    "\n",
    "multipletests_results_info.to_csv(os.path.join(home_path, \"Data\", \"FishersExactTest\", \"Fishers_multipletestcorrection_fdrbh_wd.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "23b4a3e8622309bcc6db3d5cc6eb73d60ab98d9ec23bad6a26b709981ccb403a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
